coalib/bearlib/aspects/__init__.py

```diff
@@ -0,0 +1,151 @@
+        self.example = cleandoc(example)
+        self.example_language = cleandoc(example_language)
+        self.importance_reason = cleandoc(importance_reason)
+        self.fix_suggestions = cleandoc(fix_suggestions)
+
+
+class Aspect:
+    '''
+    This is the Aspect base class. You can use it to define aspects (and sub
+    aspects recursively) including their documentation and meaningful
+    settings.
+
+    In order to create an aspect, just inherit the class from Aspect. To make
+    a subaspect, define a nested class within any aspect and inherit from it.
+    (coala will figure out which aspect belongs where via the inheritance and
+    it looks nice to use an aspect when they're defined nested.)
+
+    >>> class Redundancy(Aspect):
+    ...     docs = AspectDocumentation(
+    ...         """
+    ...         This aspect describes redundancy in your source code. Those are
+    ...         usually places where you can strip away source code in order to
+    ...         make your codebase better maintainable but keeping your full
+    ...         functionality.
+    ...         """, '...', '...', '...'
+    ...     )
+    ...
+    ...     class Clone(Aspect):
+    ...         docs = AspectDocumentation(
+    ...             """
+    ...             Code clones are different pieces of code in your codebase
+    ...             that are very similar.
+    ...             """, '...', '...',
+    ...             """
+    ...             Usually code clones can be simplified to only one
+    ...             occurrence...
+    ...             """
+    ...         )
+    ...         MIN_CLONE_TOKEN = ("The number of tokens that have to be "
+    ...                            "equal for it to be detected as a code "
+    ...                            "clone.", int, 20)
+
+    As we see in the example above, we can define settings for any aspect class
+    by providing a tuple of a documentation string, a conversion function (in
+    this case ``int`` will be applied to the given setting) and an optional
+    default value if the setting is not mandatory.
+
+    >>> Redundancy.Clone.settings  # doctest: +ELLIPSIS
+    {'min_clone_token': ('The number of tokens...', <class 'int'>, 20)}
+
+    All bears implementing a given aspect should implement all the settings
+    given and documented by it. Additional bear specific settings might be
+    added to allow further configuration.
+
+    The user can use the qualifying name according to PEP 3155 for selecting
+    and unselecting aspects (see <https://www.python.org/dev/peps/pep-3155/>):
+
+    >>> Redundancy.Clone.__qualname__
+    'Redundancy.Clone'
+    '''
+
+    docs = AspectDocumentation('', '', '', '')
+
+    def __init__(self):
+        """
+        Do not instanciate this class.
+
+        >>> Aspect()
+        Traceback (most recent call last):
+         ...
+        AssertionError: This class should not be instanciated.
+
+        Really. Don't. Ever.
```

Comment:
It's not abstract. But it shouldn't be instanciated. You're not supposed to call init ever. init of an abstract class still gets called